### ISSUE TITLE
SSE connection close automatically

### ISSUE THREAD

**Issue by Choxmi** [warmcat/libwebsockets#1486]
@lws-team Thanks again for the quick fix for the last [issue](https://github.com/warmcat/libwebsockets/issues/1485) I opened. It worked fine for the minimal-sse sample.
But when I try to integrate the same sample to my application, I face the same issue again. I'm using the same fileupload which I used in v2.0. On start it gives this lws log output. 

`[2019/01/30 13:27:49:1503] ERR: lws_h2_parse_frame_header: Failed to get ah`
`[2019/01/30 13:27:49:2780] ERR: Unknown mime-type for public_html/css/bootstrap-theme.min.css.map` 

After few seconds the count stops without any log output. It doesn't load some images as well. Can you please help me with this.
Thanks

**Comment by lws-team** [comment]
The logs are talking about something completely different than "SSE Connection close automatically".

- "failed to get ah"... have you limited the number of available ah, ie set

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-context-vhost.h#n332-338

?  If you're using h2, your browser will request everything in parallel, and lws will serve everything concurrently.  That means one page request with ten scripts / pictures / fonts needs 10 available ah to manage the 10 sub-requests.    It's best to not restrict it.

 - unknown mime type

You can teach each mount what mimetype to use for different file suffixes.  See here for an example

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-server/minimal-http-server-mimetypes/minimal-http-server-mimetypes.c#n22-36

**Comment by Choxmi** [comment]
@lws-team Thanks for the quick response.

-  Yes it was the limited number of available ah. Removed the limitation and it worked.

-  I added the mime-type and that error is also not there anymore.

But the issue remains the same. This time the connection closed without any log output. What could be the possible cause?

**Comment by lws-team** [comment]
When does it close?

There will be log output for the closing part... it might not shed any light but it'll be there... build lws with debug at cmake (-DCMAKE_BUILD_TYPE=DEBUG) then do the equivalent of -d 1039 or -d 1151 on the app (tells it to show INFO, or INFO + DEBUG + more).

The client is also able to close the connection if it feels like it.  Or if you introduce any timeout on the wsi, it will close it.  Returning nonzero from the callback will usually close it.

For some callback commands, the default action by the dummy callback is to close the connection.  Instead of ending your callback handlers with `break;` which may call the dummy callback at the end as well, it sometimes is necessary to only run your handler code and then do a `return 0;` to avoid a default action by the dummy callback that isn't what you want.

**Comment by Choxmi** [comment]
@lws-team I built lws with DEBUG and this is the output I got.
`[2019/01/30 17:17:54:0137] INFO: __lws_close_free_wsi: 0x17f33ca20: caller: timeout
[2019/01/30 17:17:54:0137] INFO: closing 0x17f33ca20: parent (nil)
[2019/01/30 17:17:54:0137] INFO:  parent 0x17f33ca20: closing children: list:
[2019/01/30 17:17:54:0137] INFO:    \---- child h2 0x17f373ec0
[2019/01/30 17:17:54:0138] INFO:    closing child 0x17f373ec0
[2019/01/30 17:17:54:0138] INFO: __lws_close_free_wsi: 0x17f373ec0: caller: h2 child recurse
[2019/01/30 17:17:54:0138] INFO:  wsi: 0x17f373ec0, his parent 0x17f33ca20: siblings:
[2019/01/30 17:17:54:0138] INFO:    \---- child h2 0x17f373ec0
[2019/01/30 17:17:54:0138] INFO: closing 0x17f373ec0: parent 0x17f33ca20
[2019/01/30 17:17:54:0138] INFO:   0x17f373ec0: disentangling from siblings
[2019/01/30 17:17:54:0138] INFO:   0x17f373ec0 disentangled from sibling (nil)
`
It looks like a timeout issue. But I have set the timeout as `lws_set_timeout(wsi, NO_PENDING_TIMEOUT, 0);`.

- I tried adding `return 0` before all the `break;`s. But still it get closed automatically. 

- I'm using my own http-server callback. That won't be a problem right?

- Can you please tell me how `The client is also able to close the connection if it feels like it.` can happen?

Again, thanks a lot for the help.

**Comment by lws-team** [comment]
When you are using h2, there is the actual network connection on the one hand and then logical "stream" wsis representing the different things going on in the single network connection.  (With h1, the network connection and the stream are the same thing).

Like the last issue, if you are carrying ws inside an h2 stream, there is special processing to notice that the h2 network connection is carrying an "immortal" stream (ws streams have no timeout by default since they are designed for long polling type behaviour) and to also let it live as long as there are immortal streams inside it.

But with SSE there's no specific upgrade of the stream like ws, it remains a normal http stream internally actually.  And lws applies timeout rules to stop h2 network connections with inactive http streams from being allowed to live forever in an idle or stalled state; if nothing is happening with any http stream inside it, the network connection is allowed to stay up for 30s, in case more streams are started.  This timeout is applied to the h2 network connection wsi (aka nwsi), not the SSE stream wsi.

And it's applied every time there's a frame on the network connection that has no h2 websocket streams.  Since there's no guarantee your SSE stream is the last stream, eg, it's still fetching pictures or fonts or whatever at the time the SSE stream is set up, you can't solve it by removing the timeout on the network wsi when the SSE stream is in LWS_CALLBACK_HTTP; another stream may cause the timeout to be reapplied later.

I pushed a patch which adds an api to mark an http stream as entering SSE mode... `lws_http_mark_sse()` and replaced the existing stuff in the minimal examples with it.  If it's http/1, then it clears the timeout and detaches any ah (so the code to do that is no longer needed, just the call to this api).  If it's http/2, then it makes the network connection act the same as it does for ws streams, ie, stops applying timeouts on it because its streams are idle.

**Comment by Choxmi** [comment]
@lws-team That's amazing. That works perfectly. Really appreciate your quick fix. Thanks

**Comment by Choxmi** [comment]
@lws-team this is not related to the issue. This is just to clarify something. Can you please tell me how to respond to a AJAX request properly in a http server. I'm using this, 
`char *dir = "/test/sample";
            p = buffer + LWS_SEND_BUFFER_PRE_PADDING;
                end = p + sizeof (buffer) - LWS_SEND_BUFFER_PRE_PADDING;
            if (!headered) {
                
                if (lws_add_http_header_status(wsi, 200, &p, end)) {
                    printf("Wrong Header\n");
                }
                if (lws_add_http_header_by_token(wsi, WSI_TOKEN_HTTP_SERVER, (unsigned char *) "paraqumpanora", 13, &p, end)) {
                    printf("Wrong Token\n");
                }
                if (lws_add_http_header_by_token(wsi, WSI_TOKEN_HTTP_CONTENT_TYPE, (unsigned char *) "application/octet-stream", 24, &p, end)) {
                    printf("Wrong Header By token\n");
                }
                if (lws_add_http_header_content_length(wsi, (int) strlen(dir), &p, end)) {
                    printf("Wrong Length\n");
                }
                if (lws_finalize_http_header(wsi, &p, end)) {
                    printf("Finalizing \n");
                }
                lws_write(wsi, buffer + LWS_SEND_BUFFER_PRE_PADDING,
                        p - (buffer + LWS_SEND_BUFFER_PRE_PADDING),
                        LWS_WRITE_HTTP_HEADERS);
                headered = true;
                lws_callback_on_writable(wsi);
                break;
            } else {
                p += lws_snprintf((char *) p, end - p,
                        "data: %s\x0d\x0a\x0d\x0a",
                        (char *) dir);
                if (lws_write(wsi, (char *) start, strlen((char *)start),
                        LWS_WRITE_HTTP) != lws_ptr_diff(p, start))
                    return 1;
                printf("Response sent\n");
                return -1;
                break;
            }`
code in `LWS_CALLBACK_HTTP_WRITEABLE` to respond to those requests. But AJAX cannot detect any response or not even an error. After some time it gives, `too much recursion` error. 

This may be off-topic. But I really appreciate your help on this. Thanks in advance.

**Comment by SunilDullolli** [comment]
@lws-team I have not limited number of available ah, i.e. max_http_header_pool = 0; even then, I get an error "failed to get ah". It seems occur when there are more simultaneous requests to load the web page data.  Does that mean I am running out of heap to create ah structs? has it got to do anything with setting this parameter: fd_limit_per_thread. Thanks

**Comment by SunilDullolli** [comment]
Is restricting max_http_header_pool a "hard" limit? I see that it easily exceeds the limit number.

**Comment by SunilDullolli** [comment]
Can someone respond to my last comment? How can I get around that?
